### PR TITLE
Converts the context passes into Sexpression transformations

### PR DESCRIPTION
**Pull request by mariari**
This Pr wishes to write the context passes for the pipeline. Also included are tests and the restructing of the code.

I tried to comment on all sections that are just moved around.

More structural changes will be needed after this PR once the ML part of the pipeline is gone. However first we need a followup PR converting the `FromFrontend` pass.

The Main new functonality is in `Juvix.Contextify.Passes` with tests being in the `Contextify` module in the test sub directory

### CODE REVIEW

**Comment by mariari** [comment]
The high code diff amount is due to moving files around, not due to actual changes.

File: library/Core/src/Juvix/Core/Common/Closure.hs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+insertGeneric :: Symbol -> T -> T
+insertGeneric name (T m) =
+  T $ Map.insert name (Info Nothing [] Nothing) m
+
+keys :: T -> Set.HashSet Symbol
+keys (T m) = Map.keysSet m
+
+lookup :: Symbol -> T -> Maybe Information
+lookup k (T m) = Map.lookup k m
```
Comment (mariari) [inline library/Core/src/Juvix/Core/Common/Closure.hs:43]: Moved Code no changes

File: library/Translate/src/Juvix/Contextify/Environment.hs [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+-- TODO ∷ make this a standard data structure
+data ErrorS
+  = CantResolve [Sexp.T]
+  | UnknownSymbol NameSymbol.T
+  | ImpossibleMoreEles
+  | Clash
+      (Shunt.Precedence Sexp.T)
+      (Shunt.Precedence Sexp.T)
+  deriving (Show, Eq)
```
Comment (mariari) [inline library/Translate/src/Juvix/Contextify/Environment.hs:34]: New Error Types

File: library/Translate/src/Juvix/Contextify/Environment.hs [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+    Just Closure.Info {} ->
+      throw @"error" (UnknownSymbol name)
+    Nothing ->
+      contextCase name
+  where
+    contextCase name =
+      case Context.lookup name ctx >>= extractInformation . Context.extractValue of
+        Nothing -> throw @"error" (UnknownSymbol name)
+        Just pr -> pure (fromMaybe Context.default' (Context.precedenceOf pr))
+
```
Comment (mariari) [inline library/Translate/src/Juvix/Contextify/Environment.hs:169]: Some new Functions

File: library/Translate/src/Juvix/Contextify/Environment.hs [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+------------------------------------------------------------
+-- searchAndClosure function dispatch table
+------------------------------------------------------------
```
Comment (mariari) [inline library/Translate/src/Juvix/Contextify/Environment.hs:172]: All Code below here is from the Last PR

**Review by mariari — COMMENTED** [review]
Outlined the parts one may skip in reviewing this

File: library/Translate/src/Juvix/Contextify/Environment.hs [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+  deriving
+    (HasThrow "error" ErrorS)
+    via MonadError MinimalAliasIO
+
+runMIO :: MinimalMIO a -> IO (Either ErrorS a, Minimal)
+runMIO (CtxIO c) = runStateT (runExceptT c) (Minimal Closure.empty)
+
+runM :: MinimalM a -> (Either ErrorS a, Minimal)
+runM (Ctx c) = runState (runExceptT c) (Minimal Closure.empty)
+
```
Comment (mariari) [inline library/Translate/src/Juvix/Contextify/Environment.hs:89]: new running environemnt

File: library/Translate/src/Juvix/Contextify/Environment.hs (hunk 2 above)
Comment (cyberglot) [reply library/Translate/src/Juvix/Contextify/Environment.hs:38]: Is that a typo? Should it be `ImpossibleMoreElse`?

File: library/Translate/src/Juvix/Contextify/Passes.hs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  (Env.ErrS m, Env.HasClosure m) => Context.T t y s -> Sexp.Atom -> Sexp.T -> m Sexp.T
+infixConversion context atom list = do
+  grouped <- groupInfix context (Sexp.Cons (Sexp.Atom atom) list)
+  case Shunt.shunt grouped of
+    Right shunted ->
+      pure $ convertShunt shunted
+    Left (Shunt.Clash pred1 pred2) ->
+      throw @"error" (Env.Clash pred1 pred2)
+    Left Shunt.MoreEles ->
+      throw @"error" Env.ImpossibleMoreEles
```
Comment (cyberglot) [inline library/Translate/src/Juvix/Contextify/Passes.hs:65]: Update here accordingly if it's a typo.

**Review by cyberglot — COMMENTED** [review]
LGTM, pls check the comment I left so I can approve it.

File: library/Translate/src/Juvix/Contextify/Environment.hs (hunk 2 above)
Comment (mariari) [reply library/Translate/src/Juvix/Contextify/Environment.hs:38]: Not a typo, it's impossible more elements

**Review by cyberglot — APPROVED** [review]

**Pull request merged**